### PR TITLE
cmd/uses: add --tree

### PR DESCRIPTION
**Pull request by SeekingMeaning**
- [ ] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Note: Unlike `brew deps`, specifying multiple formulae with `brew uses` does not show the dependents of each formula. Instead, it shows the _intersection_ of dependents of all specified formulae.

### Example usage

Dependents of `libzip` (only one level down, similar to `brew deps --1`):

```console
$ brew uses libzip --tree
libzip
├── ebook-tools
...
├── mgba
├── openrct2
├── osrm-backend
├── php
├── php@7.2
...
└── ykneomgr
```

Recursive dependents of `libzip`:

```console
$ brew uses libzip --tree --recursive
libzip
├── ebook-tools
...
├── mgba
├── openrct2
├── osrm-backend
├── php
│   ├── deployer
│   ├── easyengine
│   └── phplint
├── php@7.2
...
└── ykneomgr
```

Dependents of `libepoxy`:

```console
$ brew uses libepoxy --tree
libepoxy
├── gtk+3
├── mgba
├── nestopia-ue
└── pytouhou
```

`libzip` and `libepoxy` share a common dependent, `mgba`:

```console
$ brew uses libzip libepoxy --tree
libepoxy
libzip
└── mgba
```

---

<sup><a href="https://gist.github.com/SeekingMeaning/bf129363b32dc232e4f47d526363b733">Wonder what `brew uses openssl@1.1 --tree --recursive` looks like?</a></sup>

### CODE REVIEW

File: Library/Homebrew/cmd/uses.rb [hunk 1]
```diff
@@ -108,4 +118,80 @@ def uses
     puts Formatter.columns(uses.map(&:full_name).sort)
     odie "Missing formulae should not have dependents!" if used_formulae_missing
   end
+
+  def intersection_of_dependents(used_formulae, recursive = false)
+    if @use_runtime_dependents
+      return used_formulae.map(&:runtime_installed_formula_dependents)
+                          .reduce(&:&)
```
Comment (dawidd6) [inline Library/Homebrew/cmd/uses.rb:125]: Could you explain what this does? I'm just curious.

Comment (SeekingMeaning) [reply Library/Homebrew/cmd/uses.rb:125]: I'm not sure, I _think_ it removes duplicate entries; I copied it from the lines above

File: Library/Homebrew/cmd/uses.rb [hunk 2]
```diff
@@ -89,13 +95,13 @@ def intersection_of_dependents(use_runtime_dependents, used_formulae, args:)
...
-      deps = if args.installed?
-        dependents(Formula.installed + Cask::Caskroom.casks)
-      else
-        dependents(Formula.to_a + Cask::Cask.to_a)
-      end
+      @deps ||= (if args.installed?
+                   dependents(Formula.installed + Cask::Caskroom.casks)
+                 else
+                   dependents(Formula.to_a + Cask::Cask.to_a)
+      end).sort_by(&:full_name)
```
Comment (SeekingMeaning) [inline Library/Homebrew/cmd/uses.rb:102]:
```console
$ brew style Library/Homebrew/cmd/uses.rb

1 file inspected, no offenses detected
```

🤔

File: Library/Homebrew/cmd/uses.rb [hunk 3]
```diff
@@ -125,4 +131,44 @@ def select_used_dependents(dependents, used_formulae, recursive, includes, ignor
...
+
+  def puts_dependents_tree(use_runtime_dependents, used_formulae, args:)
+    used_formulae.sort_by(&:full_name).each do |f|
+      puts f.full_name
+    end
+    @dependents_stack = []
+    recursive_dependents_tree(use_runtime_dependents, used_formulae, "", args: args)
+  end
+
+  def recursive_dependents_tree(use_runtime_dependents, formulae, prefix, args:)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/uses.rb:143]: Could any of this logic be shared with `deps`?

Comment (SeekingMeaning) [reply Library/Homebrew/cmd/uses.rb:143]: I think so, I just need to figure out how to handle differentiating between finding the dependencies of each given formula independently (`deps`) and finding the intersection of dependents of all given formulae (`uses`)

Comment (MikeMcQuaid) [reply Library/Homebrew/cmd/uses.rb:143]: @SeekingMeaning I think I'd be more game for this functionality if we find a way of sharing this code.

Comment (SeekingMeaning) [reply Library/Homebrew/cmd/uses.rb:143]: Definitely, I just need to figure out a more abstract view of both commands. Closing this for now.

**Pull request closed without merging**